framework/src/constraints/AutomaticMortarGeneration.C

```diff
@@ -119,6 +126,25 @@ AutomaticMortarGeneration::buildMortarSegmentMesh()
   // split.
   std::map<const Elem *, std::set<Elem *>> secondary_elems_to_mortar_segments;
 
+  dof_id_type local_id_index = 0;
+  std::size_t node_unique_id_offset = 0;
+
+  // Create an offset by the maximum number of mortar segment elements that can be created *plus*
+  // the number of lower-dimensional secondary subdomain elements. Recall that the number of mortar
+  // segments created is a function of node projection, *and* that if we split elems we will delete
+  // that elem which has already taken a unique id
+  for (const auto & pr : primary_secondary_boundary_id_pairs)
+  {
+    const auto primary_bnd_id = pr.first;
+    const auto secondary_bnd_id = pr.second;
+    const auto num_primary_nodes =
+        std::distance(mesh.bid_nodes_begin(primary_bnd_id), mesh.bid_nodes_end(primary_bnd_id));
+    const auto num_secondary_nodes =
+        std::distance(mesh.bid_nodes_begin(secondary_bnd_id), mesh.bid_nodes_end(secondary_bnd_id));
+
+    node_unique_id_offset += num_primary_nodes + 2 * num_secondary_nodes;
```

Comment:
It would be great if you could have documents how you come up these magic offsets.

Reply:
I have this comment ten lines earlier:
```
  // Create an offset by the maximum number of mortar segment elements that can be created *plus*
  // the number of lower-dimensional secondary subdomain elements. Recall that the number of mortar
  // segments created is a function of node projection, *and* that if we split elems we will delete
  // that elem which has already taken a unique id
```